exercises/02_linear_search.rs

```diff
@@ -4,11 +4,11 @@ use creusot_contracts::*;
 // Prove the following:
 // 1. If we return Some(i) it is the first index containing `tgt`
 // 2. If we return None, then there are no indices containing `tgt`
-fn search<T: Ord>(v: &Vec<T>, tgt: &T) -> Option<usize> {
+fn search<T: Ord + DeepModel>(v: &[T], tgt: &T) -> Option<usize> {
```

Comment:
Are you saying compilation/translation to Coma fails without this? Or did you have to add this to prove whichever spec you wanted to prove?

Reply:
Ok, this happened after adding the specification.

```rs
use creusot_contracts::*;

#[ensures(forall<i: usize> result == Some(i) ==> i@ < v@.len() && forall<j: Int> 0 <= j && j < i@ ==> v@[j] != *tgt)]
#[ensures(result == None ==> forall<i: Int> 0 <= i && i < v@.len() ==> v@[i] != *tgt)]
pub fn search<T: Ord>(v: &[T], tgt: &T) -> Option<usize> {
    let mut i = 0;

    #[invariant(forall<j: Int> 0 <= j && j < i@ ==> v@[j] != *tgt)]
    while i < v.len() {
        if v[i] == *tgt {
            return Some(i);
        }

        i += 1
    }

    return None;
}
```

For the code above, I get the following error message:

```
error[E0277]: the trait bound `T: creusot_contracts::DeepModel` is not satisfied
  --> src/lib.rs:10:12
   |
10 |         if v[i] == *tgt {
   |            ^^^^^^^^^^^^ the trait `creusot_contracts::DeepModel` is not implemented for `T`
```

I can also use `tgt@` instead of `*tgt` in the specifications, but then I get an additional message about `ShallowModel` not being implemented by `T`.

I’m on commit `917bcee1`, if this is of interest.

Reply:
Indeed, you need to add this bound. I think my original intent was to have that be part of the exercise, but its probably best to provide that for the user.

Reply:
In fact, I started to wonder whether adding the `DeepModel` trait bound was intended to be part of the exercise once I read @sarsko’s comment :sweat_smile: I guess the actual issue is that `DeepModel` is not mentioned in the [guide](https://creusot-rs.github.io/creusot/). Otherwise you could use [`#[diagnostic::on_unimplemented]`](https://github.com/rust-lang/rust/pull/119888) and link the respective page :slightly_smiling_face: